core/src/test/java/org/elasticsearch/rest/RestHttpResponseHeadersTests.java

```diff
@@ -0,0 +1,150 @@
+     * 'Allow' header includes a list of valid HTTP methods for the endpoint
+     * (see
+     * <a href="https://tools.ietf.org/html/rfc2616#section-10.4.6">HTTP/1.1 -
+     * 10.4.6 - 405 Method Not Allowed</a>).
+     */
+    public void testUnsupportedMethodResponseHttpHeader() throws Exception {
+
+        /*
+         * Generate a random set of candidate valid HTTP methods to register
+         * with the test RestController endpoint. Enums are returned in the
+         * order they are declared, so the first step is to shuffle the HTTP
+         * method list, passing in the RandomizedContext's Random instance,
+         * before picking out a candidate sublist.
+         */
+        List<RestRequest.Method> validHttpMethodArray = new ArrayList<RestRequest.Method>(Arrays.asList(RestRequest.Method.values()));
+        Collections.shuffle(validHttpMethodArray, random());
+
+        /*
+         * The upper bound of the potential sublist is one less than the size of
+         * the array, so we are guaranteed at least one invalid method to test.
+         */
+        validHttpMethodArray = validHttpMethodArray.subList(0, randomIntBetween(1, validHttpMethodArray.size() - 1));
+        assertTrue(validHttpMethodArray.size() > 0);
+        assertTrue(validHttpMethodArray.size() < RestRequest.Method.values().length);
+
+        /*
+         * Generate an inverse list of one or more candidate invalid HTTP
+         * methods, so we have a candidate method to fire at the test endpoint.
+         */
+        List<RestRequest.Method> invalidHttpMethodArray = new ArrayList<RestRequest.Method>(Arrays.asList(RestRequest.Method.values()));
+        invalidHttpMethodArray.removeAll(validHttpMethodArray);
+        assertTrue(invalidHttpMethodArray.size() > 0);
+
+        // A basic RestHandler handles requests to the endpoint
+        RestHandler restHandler = new RestHandler() {
+            @Override
+            public void handleRequest(RestRequest request, RestChannel channel) throws Exception {
+                channel.sendResponse(new TestResponse());
+            }
+        };
+
+        // Initialize test candidate RestController
+        RestController restController = new RestController(Settings.EMPTY);
+
+        // Register valid test handlers with test RestController
+        for (RestRequest.Method method : validHttpMethodArray) {
+            restController.registerHandler(method, "/", restHandler);
+        }
+
+        // Generate a test request with an invalid HTTP method
+        RestRequest restRequest = (new FakeRestRequest() {
+
+            private RestRequest.Method method;
+
+            @Override
+            public Method method() {
+                return method;
+            }
+
+            private FakeRestRequest initialize(RestRequest.Method method) {
+                this.method = method;
+                return this;
+            }
+
+        }).initialize(invalidHttpMethodArray.get(0));
```

Comment:
Alternatively, I'm happy to add an additional constructor to the `FakeRestRequest` class accepting a `RestRequest.Method` (and default the `RestRequest.Method` to `GET`).